client/src/websockets.rs

```diff
@@ -0,0 +1,632 @@
+        Ok(self
+            .socket
+            .lock()
+            .await
+            .stream
+            .send(Message::Text(text))
+            .await?)
+    }
+
+    pub async fn send_request(
+        &mut self,
+        options: SendRequestOptions,
+    ) -> Result<(WebSocketResponseMessage)> {
+        let mut id;
+
+        //In a smaller scope to free the lock as fast as possible
+        {
+            let mut lock = self.outgoing_id.lock().await;
+            id = *lock;
+            *lock += 1;
+        }
+
+        if self.outgoing_map.lock().await.contains_key(&id) {
+            anyhow::bail!("Duplicate outgoing request")
+        }
+
+        let mut header_vec: Vec<String> = Vec::new();
+
+        if let Some(headers) = options.headers {
+            headers
+                .iter()
+                .map(|(key, value)| header_vec.push(format!("{}:{}", key, value)));
+        }
+
+        //Make the protobuf message and convert it to bytes
+        let ws_msg = WebSocketMessage {
+            r#type: Some(1),
+            request: Some(WebSocketRequestMessage {
+                id: Some(id),
+                verb: Some(options.verb),
+                path: Some(options.path),
+                headers: header_vec,
+                body: options.body,
+            }),
+            response: None,
+        };
+        let msg = Message::Binary(ws_msg.encode_to_vec());
+
+        //Send the message over ws
+        self.socket.lock().await.stream.send(msg).await?;
+
+        let mut self_clone = self.clone();
+
+        let promise = spawn(async move {
+            if self_clone.has_keepalive {
+                set_timeout(KEEPALIVE_TIMEOUT_MS.into(), || async move {
+                    self_clone
+                        .outgoing_map
+                        .lock()
+                        .await
+                        .remove(&id)
+                        .unwrap()
+                        .abort();
+                    self_clone
+                        .socket
+                        .lock()
+                        .await
+                        .stream
+                        .send(Message::Close(Some(frame::CloseFrame {
+                            code: 3008.into(),
+                            reason: "Timed out".into(),
+                        })))
+                        .await;
+                    //TODO: Kill the listener process
+                })
+                    .await;
+            }
+        });
+
+        //Add future to hashmap
+        self.outgoing_map
+            .lock()
+            .await
+            .insert(id, promise.abort_handle());
+
+        promise.await;
+
+        if let Some(incoming) = self.incoming_response.lock().await.get(&id) {
+            let status = match incoming.status {
+                Some(status) => status,
+                _ => panic!(),
+            };
+
+            if (200..300).contains(&status) {
+                Ok(incoming.clone())
```

Comment:
`use surf::{http::headers::ToHeaderValues, StatusCode};`

```suggestion
            let ok = StatusCode::Ok as u16;
            let multiple_choice = StatusCode::MultipleChoice as u16;
            if (ok..multiple_choice).contains(&(status as u16)) {
                Ok(incoming.clone())
```

Could be one-line also